### PR TITLE
Don't call __Pyx_ImportNumPyArrayTypeIfAvailable for every fused def function

### PR DESCRIPTION
**Pull request by da-woods**
Cython.Utils.OrderedSet did not have a `__bool__` function and so
was always treated as "true". The mean that Cython always generated
buffer-handling utility code (`if all_buffer_types:` in FusedNode.py)
even when just handling scalar types.

### CODE REVIEW

**Comment by da-woods** [comment]
There's no test for this because it isn't obvious how to test it. Without this change it just generates unneeded code.

File: Cython/Utils.py [hunk 1]
```diff
@@ -579,6 +579,9 @@ def add(self, e):
             self._list.append(e)
             self._set.add(e)
 
+    def __bool__(self):
+        return bool(self._set)
+
```
Comment (scoder) [inline Cython/Utils.py:585]: We still need `__nonzero__` in Py2.
```suggestion
    def __bool__(self):
        return bool(self._set)
        
    __nonzero__ = __bool__

```

File: Cython/Utils.py [hunk 2]
```diff
@@ -579,6 +579,10 @@ def add(self, e):
             self._list.append(e)
             self._set.add(e)
 
+    def __bool__(self):
+        return bool(self._set)
+        
+    __nonzero__ = __bool__
```
Comment (scoder) [inline Cython/Utils.py:585]:
```suggestion
    __nonzero__ = __bool__

```

**Comment by scoder** [comment]
> There's no test for this because it isn't obvious how to test it. Without this change it just generates unneeded code.

See https://github.com/cython/cython/pull/4938

File: tests/compile/fused_no_numpy.pyx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# cython: test_fail_if_c_code_has = __Pyx_ImportNumPyArrayTypeIfAvailable
```
Comment (da-woods) [inline tests/compile/fused_no_numpy.pyx:1]: Should we have a test for the opposite case? I don't think it's particularly useful for any except that if the fused dispatch code gets refactored then this test becomes irrelevant and we'd quite like to know that

**Comment by da-woods** [comment]
Slightly confusingly, the code style errors are in files I haven't touched

```
/home/runner/work/cython/cython/Cython/Distutils/old_build_ext.py:324: [E275] missing whitespace after keyword
runTest (__main__.TestCodeFormat) ... 
#### 2022-07-30 20:28:35.055218
/home/runner/work/cython/cython/Cython/Compiler/ExprNodes.py:547: [E275] missing whitespace after keyword
/home/runner/work/cython/cython/tests/run/pure_py.py:552: [E275] missing whitespace after keyword
/home/runner/work/cython/cython/tests/run/pure_py.py:555: [E275] missing whitespace after keyword
/home/runner/work/cython/cython/tests/run/pure_py.py:560: [E275] missing whitespace after keyword
```

**Comment by scoder** [comment]
Nice. Thanks!

**Pull request merged**